### PR TITLE
Files for 3dPFM (scripts and DEMO)

### PR DESCRIPTION
**Pull request by CesarCaballeroGaudes**

### CODE REVIEW

**Comment by zsaad** [comment]
Will do but is San Diego now. Back Thursday.
Cheers
Z

**Comment by ischwabacher** [comment]
I'm not sure how this shows up on GitHub, but I put in a PR on the PR so that it'll be included in `make install`.

I'm looking forward to playing with this program. :)

**Comment by nno** [comment]
I don't know what you mean by "I put in a PR on the PR". Did you do another commit?
I did not see another PR, and your current PR only shows a single commit, so any changes you made since the original PR are not reflected here. 

If you made a change to the code and would like to include it in this PR, please commit your changes and push to your branch (master in this case) again; github will update the PR to include the new commit(s).

**Comment by ischwabacher** [comment]
I'm not the owner of the PR, so I can't commit to it.  I forked, committed to *that*, and put in a PR on @CesarCaballeroGaudes's fork, but I was concerned (apparently correctly) that it might not show up on the grandparent tracker.

**Comment by nno** [comment]
Indeed, this PR is only for CesarCaballeroGaudes/afni:master, not CesarCaballeroGaudes/afni-1master.
If you update (i.e commit the new changes) into your afni branch (not the afni-1 branch), then it will be reflected in this PR.

**Comment by CesarCaballeroGaudes** [comment]
This might not have been the appropriate way to upload / commit the program. This is a new program we have been working on with the AFNI team (mainly Ziad)

**Comment by nno** [comment]
You made a double fork: afni/afni -> CesarCaballeroGaudes/afni, then CesarCaballeroGaudes/afni -> CesarCaballeroGaudes/afni-1.

Maybe you were trying to add new functionality without affecting your CesarCaballeroGaudes/afni master branch? If so, the preferred way is to create a new branch on CesarCaballeroGaudes/afni. Let's say that your remotes are "origin" for CesarCaballeroGaudes/afni and "upstream" for afni/afni. 
```
git checkout master # ensure we start with master branch
git pull upstream master # update master from official AFNI code
git checkout -b feature_foo # make new branch
# implement new feature
# (your work here).
#
git commit -a -m 'added new feature foo' # commit changes
git push origin feature_foo # push to your own fork on github
# now make the PR on github for this branch
git checkout master # optionally, to go back to the master branch

**Comment by ischwabacher** [comment]
It looks like the problem here was trying to do it the obvious way in the GitHub web interface, instead of doing it locally.  Let's see if this works.

**Comment by ischwabacher** [comment]
Hmm... the problem is that the afni/afni branch is refs/pull/1 and not refs/heads/something.  It looks like this can't be done with just the web interface, which is disappointing.

**Comment by nno** [comment]
@ischwabacher to me it seems best if @CesarCaballeroGaudes could update his PR (or make a new one). Preferably this would be done by creating a new branch locally (*not* a fork) and then pushing it to CesarCaballeroGaudes/afni.

**Comment by ischwabacher** [comment]
IIUC, after deleting the ischwabacher/AFNI -> afniHQ/AFNI and ischwabacher/afni-1 -> CesarCaballeroGaudes/afni forks, repopulating by forking ischwabacher/afni -> afni/afni, and repeating my makefile edits in the new fork, I currently have
* A local repository with upstream (afni/afni) and origin (ischwabacher/afni) remotes
* A remote repository ischwabacher/afni with afni/afni and CesarCaballeroGaudes/afni as remotes, with unknown names
* A branch `patch-1` in ischwabacher/afni, which is being tracked by refs/pull/2/head in Cesar/CaballeroGaudes/afni
* No ability to affect refs/pull/1/head in afni/afni.

If @CesarCaballeroGaudes merges my PR, his master -- which afni/afni's refs/pull/1/head is tracking -- will be fast-forwarded to 6b5993b, which will add that commit to this PR.  But ultimately, he has sole control over what gets added to this PR because it's tracking his master.  I could start a new PR with both his changes and mine, but that would be a mess.

So, to sum up: from the fact that that commit link works, it looks like all the GitHub forks of a given repository share the same collection of commits; different forks have different refs.  The owner of a given repository controls the branches in refs/heads, but the owner of a given pull request controls the corresponding head in refs/pull because it tracks the correct branch in the requesting fork.

**Comment by ischwabacher** [comment]
So testing this out, I get this:

```
nohup 3dPFM -input 'PASAT1_mdtwbs+tlrc[20..150]' \
            -mask /study/cfids/data/group_processed/anat/mask.nii.gz -verb 1 -outALL PFM \
            &>nohup.3dPFM.out& tail --pid=$! -f nohup.3dPFM.out
[1] 9150
nohup: ignoring input
Processing volume PASAT1_mdtwbs+tlrc[20..150] 
Loading mass library. Required for Generalized-Pseudoinverse function.
Loading abind library. Required for combining multi-dimensional arrays.
Called from: top level 
Load input dataset PASAT1_mdtwbs+tlrc[20..150] 
Loading and applying mask /study/cfids/data/group_processed/anat/mask.nii.gz to PASAT1_mdtwbs+tlrc[20..150] 
Starting computation of Paradigm Free Mapping algorithm.
Computation done: 5%: 05/14/15 13:30:03.375
Computation done: 10%: 05/14/15 13:30:20.153
Computation done: 15%: 05/14/15 13:39:08.188
Computation done: 20%: 05/14/15 14:17:40.701
Computation done: 25%: 05/14/15 15:24:50.859
Computation done: 30%: 05/14/15 17:34:43.760
Error in if (sign(bk[out_x]) == sign(ak[out_x])) { : 
  missing value where TRUE/FALSE needed
Calls: apply -> FUN -> DS_homotopy_function
In addition: There were 50 or more warnings (use warnings() to see the first 50)
Execution halted
```

It looks like there's a way to get NAs into bk or ak...  I'll do some investigating, but as you can see, it's going to take a while.

This program looks very cool, though!  The data I'm looking at right now come from a single 5 min. block of task, which is super hard to resolve with OLS... but I'm hoping I'll be able to pick out go trials or errors or something with this method.  We'll see.

**Comment by nno** [comment]
@CesarCaballeroGaudes would you have time to have a look at the issue raised by @ischwabacher, and update the PR please?

**Comment by CesarCaballeroGaudes** [comment]
@ischwabacher @noo Thanks for testing the program. Would you mind sending me your data to check why ak or bk could have NAs values for that particular voxel?

**Comment by ischwabacher** [comment]
I think that would be fine, but let me check with my PI first.

**Comment by CesarCaballeroGaudes** [comment]
These warnings don't  have any effect in the computation and just appear because the argument of min is empty. In fact, they are not shown if the program is run in command mode. Let me know about the dataset (my email address is in the code).

**Comment by ischwabacher** [comment]
We're feeling a little skittish about IRB approval of the disclosure, so when I have some time I'll go through the frame dump I was able to capture and manually create a minimal test case.

**Comment by CesarCaballeroGaudes** [comment]
I think the best is you try with the dataset of the demo (see @Install_3dPFM_Demo and the corresponding ./ccaballero/README.txt file) and see whether the program stops in your computer with that dataset. I never had fatal errors with several datasets I've analysed with the program. The Demo dataset is not in the AFNI server yet, but I can send it to you if you email me.
